File: backend/internal/queue/queue.go
package queue

import (
	"context"
	"fmt"
	"strconv"
	"time"

	"cloudfile/backend/internal/config"

	amqp "github.com/rabbitmq/amqp091-go"
)

const (
	MediaExchange     = "cloudfile.media.exchange"
	ProcessQueue      = "cloudfile.media.process.queue"
	RetryQueue        = "cloudfile.media.retry.queue"
	DeadQueue         = "cloudfile.media.dead.queue"
	ProcessRoutingKey = "media.process"
	RetryRoutingKey   = "media.retry"
	DeadRoutingKey    = "media.dead"
)

type Action string

const (
	Ack   Action = "ACK"
	Retry Action = "RETRY"
	Dead  Action = "DEAD"
)

type TaskQueue interface {
	PublishMediaTask(ctx context.Context, taskID uint) error
	PublishRetry(ctx context.Context, taskID uint) error
	PublishDead(ctx context.Context, taskID uint) error
	ConsumeMediaTask(ctx context.Context, handler func(context.Context, uint) (Action, error)) error
}

func New(cfg config.Config) (TaskQueue, error) {
	if cfg.QueueDriver == "rabbitmq" {
		return NewRabbitMQQueue(cfg.RabbitMQURL, cfg.RabbitRetryTTL)
	}
	return NewInlineQueue(), nil
}

type InlineQueue struct {
	ch chan uint
}

func NewInlineQueue() *InlineQueue {
	return &InlineQueue{ch: make(chan uint, 1024)}
}

func (q *InlineQueue) PublishMediaTask(ctx context.Context, taskID uint) error {
	select {
	case q.ch <- taskID:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

func (q *InlineQueue) PublishRetry(ctx context.Context, taskID uint) error {
	return q.PublishMediaTask(ctx, taskID)
}

func (q *InlineQueue) PublishDead(ctx context.Context, taskID uint) error {
	return nil
}

func (q *InlineQueue) ConsumeMediaTask(ctx context.Context, handler func(context.Context, uint) (Action, error)) error {
	for {
		select {
		case taskID := <-q.ch:
			_, _ = handler(ctx, taskID)
		case <-ctx.Done():
			return ctx.Err()
		default:
			return nil
		}
	}
}

type RabbitMQQueue struct {
	url      string
	retryTTL time.Duration
}

func NewRabbitMQQueue(url string, retryTTL time.Duration) (*RabbitMQQueue, error) {
	q := &RabbitMQQueue{url: url, retryTTL: retryTTL}
	conn, ch, err := q.channel()
	if err != nil {
		return nil, err
	}
	defer conn.Close()
	defer ch.Close()
	return q, q.declare(ch)
}

func (q *RabbitMQQueue) channel() (*amqp.Connection, *amqp.Channel, error) {
	conn, err := amqp.Dial(q.url)
	if err != nil {
		return nil, nil, err
	}
	ch, err := conn.Channel()
	if err != nil {
		conn.Close()
		return nil, nil, err
	}
	return conn, ch, nil
}

func (q *RabbitMQQueue) declare(ch *amqp.Channel) error {
	if err := ch.ExchangeDeclare(MediaExchange, "direct", true, false, false, false, nil); err != nil {
		return err
	}
	if _, err := ch.QueueDeclare(ProcessQueue, true, false, false, false, nil); err != nil {
		return err
	}
	if err := ch.QueueBind(ProcessQueue, ProcessRoutingKey, MediaExchange, false, nil); err != nil {
		return err
	}
	if _, err := ch.QueueDeclare(RetryQueue, true, false, false, false, amqp.Table{
		"x-message-ttl":             int32(q.retryTTL / time.Millisecond),
		"x-dead-letter-exchange":    MediaExchange,
		"x-dead-letter-routing-key": ProcessRoutingKey,
	}); err != nil {
		return err
	}
	if err := ch.QueueBind(RetryQueue, RetryRoutingKey, MediaExchange, false, nil); err != nil {
		return err
	}
	if _, err := ch.QueueDeclare(DeadQueue, true, false, false, false, nil); err != nil {
		return err
	}
	return ch.QueueBind(DeadQueue, DeadRoutingKey, MediaExchange, false, nil)
}

func (q *RabbitMQQueue) publish(ctx context.Context, routingKey string, taskID uint) error {
	conn, ch, err := q.channel()
	if err != nil {
		return err
	}
	defer conn.Close()
	defer ch.Close()
	if err := q.declare(ch); err != nil {
		return err
	}
	if err := ch.Confirm(false); err != nil {
		return err
	}
	confirms := ch.NotifyPublish(make(chan amqp.Confirmation, 1))
	if err := ch.PublishWithContext(ctx, MediaExchange, routingKey, true, false, amqp.Publishing{
		ContentType:  "text/plain",
		DeliveryMode: amqp.Persistent,
		Body:         []byte(strconv.FormatUint(uint64(taskID), 10)),
	}); err != nil {
		return err
	}
	select {
	case confirmation := <-confirms:
		if !confirmation.Ack {
			return fmt.Errorf("rabbitmq publish was not confirmed")
		}
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

func (q *RabbitMQQueue) PublishMediaTask(ctx context.Context, taskID uint) error {
	return q.publish(ctx, ProcessRoutingKey, taskID)
}

func (q *RabbitMQQueue) PublishRetry(ctx context.Context, taskID uint) error {
	return q.publish(ctx, RetryRoutingKey, taskID)
}

func (q *RabbitMQQueue) PublishDead(ctx context.Context, taskID uint) error {
	return q.publish(ctx, DeadRoutingKey, taskID)
}

func (q *RabbitMQQueue) ConsumeMediaTask(ctx context.Context, handler func(context.Context, uint) (Action, error)) error {
	conn, ch, err := q.channel()
	if err != nil {
		return err
	}
	defer conn.Close()
	defer ch.Close()
	if err := q.declare(ch); err != nil {
		return err
	}
	if err := ch.Qos(1, 0, false); err != nil {
		return err
	}
	deliveries, err := ch.Consume(ProcessQueue, "", false, false, false, false, nil)
	if err != nil {
		return err
	}
	for {
		select {
		case delivery := <-deliveries:
			taskID64, err := strconv.ParseUint(string(delivery.Body), 10, 64)
			if err != nil {
				_ = delivery.Ack(false)
				continue
			}
			action, err := handler(ctx, uint(taskID64))
			if err == nil {
				switch action {
				case Retry:
					err = q.PublishRetry(ctx, uint(taskID64))
				case Dead:
					err = q.PublishDead(ctx, uint(taskID64))
				}
			}
			if err != nil {
				_ = delivery.Nack(false, true)
				continue
			}
			_ = delivery.Ack(false)
		case <-ctx.Done():
			return ctx.Err()
		}
	}
}

File: cloudfile/queue.py
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from enum import StrEnum

import pika


MEDIA_EXCHANGE = "cloudfile.media.exchange"
PROCESS_QUEUE = "cloudfile.media.process.queue"
RETRY_QUEUE = "cloudfile.media.retry.queue"
DEAD_QUEUE = "cloudfile.media.dead.queue"
PROCESS_ROUTING_KEY = "media.process"
RETRY_ROUTING_KEY = "media.retry"
DEAD_ROUTING_KEY = "media.dead"


@dataclass(frozen=True)
class MediaTaskMessage:
    task_id: int


class QueueAction(StrEnum):
    ACK = "ACK"
    RETRY = "RETRY"
    DEAD = "DEAD"


class TaskQueue:
    def publish_media_task(self, message: MediaTaskMessage) -> None:
        raise NotImplementedError

    def consume_media_task(self, handler: Callable[[MediaTaskMessage], QueueAction | None]) -> None:
        raise NotImplementedError

    def publish_retry(self, message: MediaTaskMessage) -> None:
        self.publish_media_task(message)

    def publish_dead(self, message: MediaTaskMessage) -> None:
        self.publish_media_task(message)

    def publish(self, message: MediaTaskMessage) -> None:
        self.publish_media_task(message)


class InlineQueue(TaskQueue):
    def __init__(self) -> None:
        self.messages: list[MediaTaskMessage] = []
        self.retry_messages: list[MediaTaskMessage] = []
        self.dead_messages: list[MediaTaskMessage] = []

    def publish_media_task(self, message: MediaTaskMessage) -> None:
        self.messages.append(message)

    def publish_retry(self, message: MediaTaskMessage) -> None:
        self.retry_messages.append(message)
        self.messages.append(message)

    def publish_dead(self, message: MediaTaskMessage) -> None:
        self.dead_messages.append(message)

    def consume_media_task(self, handler: Callable[[MediaTaskMessage], QueueAction | None]) -> None:
        while self.messages:
            message = self.messages.pop(0)
            action = handler(message) or QueueAction.ACK
            if action == QueueAction.RETRY:
                self.publish_retry(message)
            elif action == QueueAction.DEAD:
                self.publish_dead(message)


class RabbitMQQueue(TaskQueue):
    def __init__(self, url: str, retry_ttl_ms: int = 30_000) -> None:
        self.url = url
        self.retry_ttl_ms = retry_ttl_ms
        self._setup()

    def _connect(self):
        return pika.BlockingConnection(pika.URLParameters(self.url))

    def _setup_channel(self, channel) -> None:
        channel.exchange_declare(exchange=MEDIA_EXCHANGE, exchange_type="direct", durable=True)
        channel.queue_declare(queue=PROCESS_QUEUE, durable=True)
        channel.queue_bind(queue=PROCESS_QUEUE, exchange=MEDIA_EXCHANGE, routing_key=PROCESS_ROUTING_KEY)
        channel.queue_declare(
            queue=RETRY_QUEUE,
            durable=True,
            arguments={
                "x-message-ttl": self.retry_ttl_ms,
                "x-dead-letter-exchange": MEDIA_EXCHANGE,
                "x-dead-letter-routing-key": PROCESS_ROUTING_KEY,
            },
        )
        channel.queue_bind(queue=RETRY_QUEUE, exchange=MEDIA_EXCHANGE, routing_key=RETRY_ROUTING_KEY)
        channel.queue_declare(queue=DEAD_QUEUE, durable=True)
        channel.queue_bind(queue=DEAD_QUEUE, exchange=MEDIA_EXCHANGE, routing_key=DEAD_ROUTING_KEY)

    def _setup(self) -> None:
        connection = self._connect()
        try:
            channel = connection.channel()
            self._setup_channel(channel)
        finally:
            connection.close()

    def _publish(self, routing_key: str, message: MediaTaskMessage) -> None:
        connection = self._connect()
        try:
            channel = connection.channel()
            self._setup_channel(channel)
            channel.confirm_delivery()
            channel.basic_publish(
                exchange=MEDIA_EXCHANGE,
                routing_key=routing_key,
                body=str(message.task_id).encode(),
                mandatory=True,
                properties=pika.BasicProperties(delivery_mode=pika.DeliveryMode.Persistent),
            )
        finally:
            connection.close()

    def publish_media_task(self, message: MediaTaskMessage) -> None:
        self._publish(PROCESS_ROUTING_KEY, message)

    def publish_retry(self, message: MediaTaskMessage) -> None:
        self._publish(RETRY_ROUTING_KEY, message)

    def publish_dead(self, message: MediaTaskMessage) -> None:
        self._publish(DEAD_ROUTING_KEY, message)

    def consume_media_task(self, handler: Callable[[MediaTaskMessage], QueueAction | None]) -> None:
        connection = self._connect()
        channel = connection.channel()
        self._setup_channel(channel)
        channel.basic_qos(prefetch_count=1)

        def callback(ch, method, properties, body):
            message = MediaTaskMessage(task_id=int(body.decode()))
            action = handler(message) or QueueAction.ACK
            if action == QueueAction.RETRY:
                self.publish_retry(message)
            elif action == QueueAction.DEAD:
                self.publish_dead(message)
            ch.basic_ack(delivery_tag=method.delivery_tag)

        channel.basic_consume(queue=PROCESS_QUEUE, on_message_callback=callback, auto_ack=False)
        channel.start_consuming()


MediaQueue = TaskQueue
InlineMediaQueue = InlineQueue
RabbitMQMediaQueue = RabbitMQQueue
